lib/srv/db/cassandra/engine.go

```diff
@@ -0,0 +1,312 @@
+	// clientConn is a client connection.
+	clientConn *protocol.Conn
+	// sessionCtx is current session context.
+	sessionCtx *common.Session
+	// handshakeTriggered is set to true if handshake was triggered and
+	// used to indicated that custom errors should be sent to the client.
+	// Cassandra wire protocol relays on streamID to that needs to match the request value
+	// so sending a custom error to the client requires to read a first message send by the client.
+	handshakeTriggered bool
+}
+
+// SendError send a Cassandra ServerError to  error to the client if handshake was not yet initialized by the client.
+// Cassandra wire protocol relays on streamID to that are set by the client and server response needs to
+// set the correct streamID in order to get streamID SendError reads a first message send by the client.
+func (e *Engine) SendError(sErr error) {
+	if utils.IsOKNetworkError(sErr) || sErr == nil {
+		return
+	}
+	e.Log.Debug("cassandra connection error: %v", sErr)
+	if e.handshakeTriggered {
+		return
+	}
+
+	eh := failedHandshake{error: sErr}
+	if err := eh.handshake(e.clientConn, nil); err != nil {
+		e.Log.Warnf("cassandra connection error: %v", sErr)
+	}
+}
+
+// InitializeConnection initializes the database connection.
+func (e *Engine) InitializeConnection(clientConn net.Conn, sessionCtx *common.Session) error {
+	e.clientConn = protocol.NewConn(clientConn)
+	e.sessionCtx = sessionCtx
+	return nil
+}
+
+// HandleConnection processes the connection from Cassandra proxy coming
+// over reverse tunnel.
+func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Session) error {
+	err := e.authorizeConnection(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	serverConn, err := e.connect(ctx, sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer serverConn.Close()
+
+	e.Audit.OnSessionStart(e.Context, sessionCtx, nil)
+	defer e.Audit.OnSessionEnd(e.Context, sessionCtx)
+
+	if err := e.handshake(sessionCtx, e.clientConn, serverConn); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return trace.Wrap(e.handleClientServerConn(ctx, e.clientConn, serverConn))
+}
+
+func (e *Engine) handleClientServerConn(ctx context.Context, clientConn *protocol.Conn, serverConn net.Conn) error {
+	errC := make(chan error, 2)
+	go func() {
+		err := e.handleClientConnectionWithAudit(clientConn, serverConn)
+		errC <- trace.Wrap(err, "client done")
+	}()
+	go func() {
+		err := e.handleServerConnection(clientConn, serverConn)
+		errC <- trace.Wrap(err, "server done")
+	}()
+
+	var errs []error
+	for i := 0; i < 2; i++ {
+		select {
+		case <-ctx.Done():
+			return trace.Wrap(ctx.Err())
+		case err := <-errC:
+			if err != nil && !utils.IsOKNetworkError(errors.Unwrap(err)) && !errors.Is(err, io.EOF) {
+				errs = append(errs, err)
+			}
+		}
+	}
+	return trace.NewAggregate(errs...)
+
+}
+
+func (e *Engine) handleClientConnectionWithAudit(clientConn *protocol.Conn, serverConn net.Conn) error {
+	defer serverConn.Close()
```

Comment:
It would appear `serverConn.Close()` is called twice: here and in `HandleConnection()`? 

I think the latter is more natural; perhaps it should handle the client connection too?

Reply:
This is our patter used during proxy'ing data between two connection. When the `handleClientConnectionWithAudit` exist it needs to close the  `serverConn` to trigger the  `io.Copy(client, server)` to exit. Where the `serverConn.Close()` is `HandleConnection` is called to before handler can exit before the proxy logic is triggered. Calling conn.Close() is safe and won't case any issue.